RecoTracker/MkFitCore/interface/Track.h

```diff
@@ -0,0 +1,800 @@
+
+    int nUniqueLayers() const {
+      // make local copy in vector: sort it in place
+      std::vector<HitOnTrack> tmp_hitsOnTrk(hitsOnTrk_.begin(), hitsOnTrk_.end());
+      std::sort(tmp_hitsOnTrk.begin(), tmp_hitsOnTrk.end(), [](const auto& h1, const auto& h2) {
+        return h1.layer < h2.layer;
+      });
+
+      // local counters
+      auto lyr_cnt = 0;
+      auto prev_lyr = -1;
+
+      // loop over copy of hitsOnTrk
+      for (auto ihit = 0; ihit <= lastHitIdx_; ++ihit) {
+        const auto& hot = tmp_hitsOnTrk[ihit];
+        const auto lyr = hot.layer;
+        const auto idx = hot.index;
+        if (lyr >= 0 && (idx >= 0 || idx == -9) && lyr != prev_lyr) {
+          ++lyr_cnt;
+          prev_lyr = lyr;
+        }
+      }
+      return lyr_cnt;
+    }
+
+    // this method sorts the data member hitOnTrk_ and is ONLY to be used by sim track seeding
+    void sortHitsByLayer();
+
+    // used by fittest only (NOT mplex)
+    const std::vector<int> foundLayers() const {
+      std::vector<int> layers;
+      for (int ihit = 0; ihit <= lastHitIdx_; ++ihit) {
+        if (hitsOnTrk_[ihit].index >= 0 || hitsOnTrk_[ihit].index == -9) {
+          layers.push_back(hitsOnTrk_[ihit].layer);
+        }
+      }
+      return layers;
+    }
+
+    Track clone() const { return Track(*this); }
+
+  private:
+    std::vector<HitOnTrack> hitsOnTrk_;
+  };
+
+  typedef std::vector<Track> TrackVec;
+  typedef std::vector<TrackVec> TrackVecVec;
+
+  inline bool sortByHitsChi2(const Track& cand1, const Track& cand2) {
+    if (cand1.nFoundHits() == cand2.nFoundHits())
+      return cand1.chi2() < cand2.chi2();
+    return cand1.nFoundHits() > cand2.nFoundHits();
+  }
+
+  inline bool sortByScoreCand(const Track& cand1, const Track& cand2) { return cand1.score() > cand2.score(); }
+
+  inline bool sortByScoreStruct(const IdxChi2List& cand1, const IdxChi2List& cand2) {
+    return cand1.score > cand2.score;
+  }
+
+  inline float getScoreWorstPossible() {
+    return -1e16;  // somewhat arbitrary value, used for handling of best short track during finding (will try to take it out)
+  }
+
+  inline float getScoreCalc(const int nfoundhits,
+                            const int ntailholes,
+                            const int noverlaphits,
+                            const int nmisshits,
+                            const float chi2,
+                            const float pt,
+                            const bool inFindCandidates = false) {
+    //// Do not allow for chi2<0 in score calculation
+    // if(chi2<0) chi2=0.f;
+
+    float maxBonus = 8.0;
+    float bonus = Config::validHitSlope_ * nfoundhits + Config::validHitBonus_;
+    float penalty = Config::missingHitPenalty_;
+    float tailPenalty = Config::tailMissingHitPenalty_;
+    float overlapBonus = Config::overlapHitBonus_;
+    if (pt < 0.9) {
+      penalty *= inFindCandidates ? 1.7f : 1.5f;
+      bonus = std::min(bonus * (inFindCandidates ? 0.9f : 1.0f), maxBonus);
+    }
```

Comment:
shiould the pt cut and the related stuff be in a config?

Reply:
This is somewhat ad hoc free-form optimization, so, it's better to be hardcoded (to avoid having to implement support for free form functionals).
Having said that, there is a somewhat near future plan to allow variations in ranking for different iterations. The configurability will be resolved/implemented in a more straightforward way then.